[AbiHalim] iP

AbiHalim opened the pull request:
# Joni
> Hello! My name is Joni. And this is my promise: helping you!

Joni is a comprehensive all-in-one to-do list companion to help you remember what you need to do. It's
- Text-based
- Easy to learn
- ~~FAST~~ _SUPER_ FAST to use

## Why Joni?
Joni takes on the persona of a helpful and kind companion, inspired by the character Joni from the 2005 Indonesian Film [_Janji Joni_ ](https://www.youtube.com/watch?v=Fe6-TV79Z4k) _(Joni's Promise)_. 

## What's Joni?
Joni allows you to record all the tasks, events, and deadlines you have, for example:
- [x] Finish CS2103 iP PR review
- [ ] Get dinner at Supper Stretch
- [ ] Buy laundry detergent

## Features:
1. list - Shows all tasks.
2.  todo <description> - Adds a new todo task.
3.  deadline <description> /by <date> - Adds a deadline task.
4.  event <description> /from <date> /to <date> - Adds an event.
5.  mark <task number> - Marks a task as completed.
6.  unmark <task number> - Marks a task as not done.
7.  delete <task number> - Removes a task.
8.  help - Displays this help message.
9.  find <keyword> - Finds all tasks containing <keyword>  10. bye - Exits the chatbot.

If you are interested in programming 🤓, here is how the above 'help' command is implemented:
```bash
package joni.command;

import joni.task.TaskList;
import joni.Ui;

/**
 * Displays available commands.
 */
public class HelpCommand extends Command {

    /**
     * Displays a list of available commands.
     *
     * @param tasks The TaskList instance.
     * @param ui The UI instance to display messages.
     */
    @Override
    public void execute(TaskList tasks, Ui ui) {
        String helpMessage = "Here are the available commands:\n"
                + " 1. list - Shows all tasks.\n"
                + " 2. todo <description> - Adds a new todo task.\n"
                + " 3. deadline <description> /by <date> - Adds a deadline task.\n"
                + " 4. event <description> /from <date> /to <date> - Adds an event.\n"
                + " 5. mark <task number> - Marks a task as completed.\n"
                + " 6. unmark <task number> - Marks a task as not done.\n"
                + " 7. delete <task number> - Removes a task.\n"
                + " 8. help - Displays this help message.\n"
                + " 9. find <keyword> - Finds all tasks containing <keyword> "
                + " 10. bye - Exits the chatbot.";
        ui.printMessage(helpMessage);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

AbiHalim committed:
Renamed Duke to Joni

AbiHalim committed:
Implemented Echo

AbiHalim committed:
Added task list functionality

AbiHalim committed:
Added feature to mark as done, task class

AbiHalim committed:
ToDos, Events, Deadlines

AbiHalim committed:
Automated Text UI Testing

AbiHalim committed:
Added error handling

AbiHalim committed:
Added DeleteTask functionality

AbiHalim committed:
Implemented Enum and updated Automated Testing

AbiHalim committed:
Updated iP.AI log

AbiHalim committed:
Refactor directory and implement rudimentary storage feature

AbiHalim committed:
Fix switch case indentation in CommandType.java

AbiHalim committed:
Add help command

AbiHalim committed:
Fix bug with help command and changed some responses

AbiHalim committed:
Merge branch 'branch-level-7'

AbiHalim committed:
Add LocalDate support

AbiHalim committed:
Fix bugs with parsing dates

AbiHalim committed:
Update savedata.csv

AbiHalim committed:
Merge branch 'branch-level-8'

AbiHalim committed:
Refactor code into more classes

AbiHalim committed:
Created new class for all tasks

AbiHalim committed:
Merge branch 'more-oop'

AbiHalim committed:
Merge remote-tracking branch 'origin/add-gradle-support'

AbiHalim committed:
Update build.gradle

AbiHalim committed:
Add JUnit tests

AbiHalim committed:
Update AI.md

AbiHalim committed:
Merge branch 'gradle-junit-jar'

AbiHalim committed:
Update AI.md

AbiHalim committed:
Add JavaDoc comments

AbiHalim committed:
Tweak code to comply with coding standard

I could not find that many violations (as far as I can tell) except I changed some names of final variables to all caps

AbiHalim committed:
Implement Find command

AbiHalim committed:
Refactor the command package to include Parser and CommandType

AbiHalim committed:
Merge branch 'A-CodingStandard-Fixed'

AbiHalim committed:
Merge branch 'branch-Level-9'

AbiHalim committed:
Merge branch 'branch-A-JavaDoc'

YosieSYX commented on src/main/java/joni/Storage.java:
```diff
@@ -0,0 +1,114 @@
+        try (BufferedReader reader = new BufferedReader(new FileReader(FILE_PATH))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                Task task = parseTaskFromCsv(line);
```
I like this meaningful function name.

YosieSYX commented on src/main/java/joni/command/Command.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Abstract Command class that all commands inherit from.
+ */
+public abstract class Command {
```
I like how you created a abstract class Command and let other classes extend this. It makes your code more organized.

YosieSYX commented on src/main/java/joni/command/HelpCommand.java:
```diff
@@ -0,0 +1,32 @@
+     * @param ui The UI instance to display messages.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui) {
```
I like this extra feature of "help" command that you implemented.

YosieSYX commented on src/main/java/joni/task/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+     * @return A string representation of the task in CSV format.
+     */
+    @Override
+    public String toCsvFormat() {
```
Would it be better to start this function with a verb, e.g. name it like: convertToCsvFormat()?

YosieSYX commented on src/test/java/joni/JoniTest.java:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Tests for the Joni chatbot.
+ */
+public class JoniTest {
```
I like the meaningful function names for test cases.

YosieSYX reviewed (commented):
LGTM. I feel the overall code is clean and organised. A minor code standard violation that I noticed is that some of your function names did not start with a verb. So maybe consider adjusting some function names.

CloudKai commented on src/main/java/joni/command/CommandType.java:
```diff
@@ -0,0 +1,22 @@
+package joni.command;
+
+enum CommandType {
+    BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, UNKNOWN, HELP, FIND;
+
+    public static CommandType fromString(String command) {
+        switch (command.toLowerCase()) {
+        case "help": return HELP;
+        case "bye": return BYE;
+        case "exit": return BYE;
+        case "list": return LIST;
+        case "mark": return MARK;
+        case "unmark": return UNMARK;
+        case "todo": return TODO;
+        case "deadline": return DEADLINE;
+        case "event": return EVENT;
+        case "delete": return DELETE;
+        case "find": return FIND;
+        default: return UNKNOWN;
```
Good use of enum for the command types, I also like how you separate the functions to keep the code looking clean

CloudKai commented on src/test/java/joni/ParserTest.java:
```diff
@@ -0,0 +1,78 @@
+        assertTrue(command instanceof ListCommand);
+    }
+
+    @Test
+    public void parse_validDeleteCommand_returnsDeleteCommand() throws JoniException {
+        Command command = Parser.parse("delete 2");
+        assertTrue(command instanceof DeleteCommand);
+    }
+
+    @Test
+    public void parse_validMarkCommand_returnsMarkCommand() throws JoniException {
+        Command command = Parser.parse("mark 1");
+        assertTrue(command instanceof MarkCommand);
+    }
+
+    @Test
+    public void parse_validUnmarkCommand_returnsMarkCommand() throws JoniException {
+        Command command = Parser.parse("unmark 1");
+        assertTrue(command instanceof MarkCommand);
+    }
+
+    @Test
+    public void parse_validHelpCommand_returnsHelpCommand() throws JoniException {
+        Command command = Parser.parse("help");
+        assertTrue(command instanceof HelpCommand);
+    }
+
+    @Test
+    public void parse_validExitCommand_returnsExitCommand() throws JoniException {
+        Command command = Parser.parse("bye");
+        assertTrue(command instanceof ExitCommand);
+    }
+
+    @Test
+    public void parse_invalidCommand_throwsJoniException() {
+        Exception exception = assertThrows(JoniException.class, () -> {
+            Parser.parse("invalidcommand");
+        });
+        assertEquals("Oops! I don't understand that command.", exception.getMessage());
+    }
```
Good tests you did here very detailed and clean

CloudKai commented on src/main/java/joni/task/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+
+        StringBuilder sb = new StringBuilder("Here are the tasks in your list:\n");
+        for (int i = 0; i < tasks.size(); i++) {
+            sb.append(" ").append(i + 1).append(". ").append(tasks.get(i)).append("\n");
```
Any reason why you did it this way instead of that way?
sb.append(String.format(" %d. %s%n", i + 1, tasks.get(i)));

CloudKai commented on src/main/java/joni/Ui.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Displays the welcome message when the chatbot starts.
+     */
+    public void showWelcome() {
+        System.out.println(DIVIDER);
+        String logo = "    .---.    .-'''-.                    \n" +
+                "    |   |   '   _    \\                  \n" +
+                "    '---' /   /` '.   \\    _..._   .--. \n" +
+                "    .---..   |     \\  '  .'     '. |__| \n" +
+                "    |   ||   '      |  '.   .-.   ..--. \n" +
+                "    |   |\\    \\     / / |  '   '  ||  | \n" +
+                "    |   | `.   ` ..' /  |  |   |  ||  | \n" +
+                "    |   |    '-...-'`   |  |   |  ||  | \n" +
+                "    |   |               |  |   |  ||  | \n" +
+                "    |   |               |  |   |  ||__| \n" +
+                " __.'   '               |  |   |  |     \n" +
+                "|      '                |  |   |  |     \n" +
+                "|____.'                 '--'   '--'     \n";
+        System.out.println(logo);
+        System.out.println(" Hello! My name is Joni");
+        System.out.println(" And this is my promise: helping you!");
+        System.out.println(" Type \"help\" for a list of commands.");
+        System.out.println(DIVIDER);
+    }
```
I like the logo you made here made the program more personal feels

CloudKai reviewed (commented):
The code is neat and looks good.
It was a pleasure to review your PR

AbiHalim committed:
Update README.md

AbiHalim committed:
Fix checkstyle violations

AbiHalim committed:
Merge branch 'checkstyle'

AbiHalim committed:
Create GUI

Version 1

AbiHalim committed:
Fix bugs

AbiHalim committed:
Fix style issues

AbiHalim committed:
Merge branch 'GUI'

AbiHalim committed:
Update README.md

Dummy branch because I named the Level-10 branch the wrong name previously (GUI)

AbiHalim committed:
Merge branch 'branch-Level-10'

AbiHalim committed:
Update AI.md

AbiHalim committed:
Update AI.md

ChauuuLe commented on src/main/java/joni/Main.java:
```diff
@@ -0,0 +1,36 @@
+     */
+    @Override
+    public void start(Stage stage) {
+        try {
```
Consider breaking it into helper methods for scalability

ChauuuLe reviewed (commented):
Almost no issues

OvOtter commented on src/main/java/joni/Storage.java:
```diff
@@ -0,0 +1,122 @@
+    private static Task parseTaskFromCsv(String csvLine) {
+        String[] parts = csvLine.split(", ");
+        if (parts.length < 3) {
+            System.out.println("Corrupted data detected: " + csvLine);
```
This can be abstracted into a method in Ui.java for a stronger SLAP.

OvOtter reviewed (commented):
LGTM in overall! Just a minor advice on code quality.

AbiHalim committed:
Merge branch 'master' of https://github.com/AbiHalim/ip

AbiHalim committed:
Improve code style

Improve code style

My peers pointed out some points for improvement in my code in the PR review.

The improvements will improve the readability of the code.

Let's address some of the suggested improvements throughout the codebase

AbiHalim committed:
Fix more issues in code quality

AbiHalim committed:
Remove text-based UI

Remove text-based UI

The app runs on a graphical user interface and no longer uses the text-based UI, however the text-based UI files are still there

The presence of text-based UI files take up extra memory and may be confusing

Removing the text-based UI in Ui.java, as well as the automated text UI testing from the codebase allows us to free up space and make it less confusing

Let's remove them from the codebase

AbiHalim committed:
Add assertions

AbiHalim committed:
Merge pull request #1 from AbiHalim/branch-A-CodeQuality

Improve code quality

AbiHalim committed:
Merge branch 'branch-A-Assertions' into master

AbiHalim committed:
Merge pull request #2 from AbiHalim/master

Merge master to Assertions

AbiHalim committed:
Merge pull request #3 from AbiHalim/branch-A-Assertions

Merge pull request #2 from AbiHalim/master

AbiHalim committed:
Minor bug fix

AbiHalim committed:
Add undo command

AbiHalim committed:
Merge branch 'branch-C-Undo'

AbiHalim committed:
Fix minor bugs

AbiHalim committed:
Fix 'Event' command

AbiHalim committed:
Change profile images

AbiHalim committed:
Implement undo of marking and unmarking tasks

AbiHalim committed:
Update README.md

AbiHalim committed:
Update acknowledgement on AI use

AbiHalim committed:
Refactor GUI related classes into new package

AbiHalim committed:
Update JUnit tests

AbiHalim committed:
Add product screenshot

AbiHalim committed:
Update README.md

AbiHalim committed:
Update README.md

AbiHalim committed:
Merge branch 'final-release'

AbiHalim committed:
Add Windows and Linux JavaFX dependencies
